### PR TITLE
Doc: Perlmutter Python & PSATD+RZ Environment

### PR DESCRIPTION
**Pull request by ax3l**
Extent the Perlmutter (NERSC) docs to install also the Python bindings.
Also install BLAS++ & LAPACK++ for PSATD+RZ support.

Summit: improve comments for the same workflows.

Perlmutter/Cori (NERSC): Add More Jupyter Modules for DASK workflows with openPMD-api.

### CODE REVIEW

File: Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example [hunk 1]
```diff
@@ -9,19 +9,33 @@ module load cudatoolkit
...
+  source $HOME/sw/perlmutter/venvs/warpx/bin/activate
+fi
 
 # optional: an alias to request an interactive node for two hours
 function getNode() {
     salloc -N 1 --ntasks-per-node=4 -t 2:00:00 -C gpu -c 32 -G 4 -A $proj
 }
 
+# an alias to request an interactive batch node for one hour
+#   for paralle execution, start on the batch node: srun <command>
```
Comment (roelof-groenewald) [inline Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example]:
```suggestion
#   for parallel execution, start on the batch node: srun <command>
```

**Comment by roelof-groenewald** [comment]
I'm getting an error:
`MPIDI_CRAY_init: GPU_SUPPORT_ENABLED is requested, but GTL library is not linked`
when trying to launch a simulation after following the above workflow. The NERSC documentation describes this error:
https://docs.nersc.gov/development/compilers/wrappers/#set-the-accelerator-target-to-gpus-for-cuda-aware-mpi-on-perlmutter

**Comment by ax3l** [comment]
Hm, we already set the right variable: #2736 

Can you check if `mpi4py` shows that problem or if a C++ version of WarpX has the same problem for you?

**Comment by roelof-groenewald** [comment]
> Hm, we already set the right variable: #2736
> 
> Can you check if `mpi4py` shows that problem or if a C++ version of WarpX has the same problem for you?

Will do!

**Comment by roelof-groenewald** [comment]
@ax3l I did a fresh install of c-blosc, ADIOS2, BLAS++, LAPACK++ and WarpX while definitely having the environment variable `CRAY_ACCEL_TARGET=nvidia80` set. The WarpX binary executes fine running a C++ only simulation. But unfortunately I get the same error as before when trying to execute a PICMI script. The error occurs when calling `import pywarpx` but not when calling `import mpi4py`.

**Comment by ax3l** [comment]
Thanks, can reproduce! I'll take a look where this is coming from :)

**Comment by ax3l** [comment]
Ah, can reproduce now. It's definitely `mpi4py`:
```
#SBATCH -t 00:05:00
#SBATCH -N 1
#SBATCH -J mpi4py
#SBATCH -A m3906_g
#SBATCH -q early_science
#SBATCH -C gpu
#SBATCH -c 32
#SBATCH --ntasks-per-node=4
#SBATCH --gpus-per-task=1
#SBATCH --gpu-bind=single:1
#SBATCH -o mpi4py.o%j
#SBATCH -e mpi4py.e%j

# ============
# -N =                 nodes
# -n =                 tasks (MPI ranks, usually = G)
# -G =                 GPUs (full Perlmutter node, 4)
# -c =                 CPU per task (128 total threads on CPU, 32 per GPU)
#
# --ntasks-per-node=   number of tasks (MPI ranks) per node (full node, 4)
# --gpus-per-task=     number of GPUs per task (MPI rank) (full node, 4)
# --gpus-per-node=     number of GPUs per node (full node, 4)
#
# --gpu-bind=single:1  sets only one GPU to be visible to each MPI rank
#                         (quiets AMReX init warnings)
#
# Recommend using --ntasks-per-node=4, --gpus-per-task=1 and --gpu-bind=single:1,
# as they are fixed values and allow for easy scaling with less adjustments.
#
# ============

# GPU-aware MPI
export MPICH_GPU_SUPPORT_ENABLED=1

EXE=./warpx
#EXE=../WarpX/build/bin/warpx.3d.MPI.CUDA.DP.OPMD.QED
#EXE=./main3d.gnu.TPROF.MPI.CUDA.ex
INPUTS=inputs_small

prg="from mpi4py import MPI
import sys

size = MPI.COMM_WORLD.Get_size()
rank = MPI.COMM_WORLD.Get_rank()
name = MPI.Get_processor_name()

sys.stdout.write(
    'Hello, World! I am process %d of %d on %s.\n'
    % (rank, size, name))"

srun python -c "$prg"
```
```
MPICH ERROR [Rank 0] [job id 1416014.0] [Fri Feb 11 19:03:11 2022] [nid002209] - Abort(-1) (rank 0 in comm 0): MPIDI_CRAY_init: GPU_SUPPORT_ENABLED is requested, but GTL library is not linked
 (Other MPI error)

aborting job:
MPIDI_CRAY_init: GPU_SUPPORT_ENABLED is requested, but GTL library is not linked

srun: error: nid002209: task 0: Exited with exit code 255
srun: launch/slurm: _step_signal: Terminating StepId=1416014.0
srun: error: nid002209: tasks 1-3: Segmentation fault
```

Will tweak its install.

**Comment by ax3l** [comment]
Ok, I fixed the `mpi4py` instructions: https://docs.nersc.gov/development/languages/python/using-python-perlmutter/#building-cuda-aware-mpi4py

Do this to fix your current env:
```bash
python3 -m pip uninstall mpi4py
python3 -m pip cache purge

MPICC="cc -target-accel=nvidia80 -shared" python3 -m pip install --upgrade mpi4py --no-build-isolation --no-binary mpi4py
```

For the `mpi4py` setuptools logic, it seems to ignore the env variable `CRAY_ACCEL_TARGET` even if we build w/o build isolation, thus we have to pass in the flags manually.

**Comment by ax3l** [comment]
I think I will add proper RPATH handling now, too. About time. Separate PR.

File: Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example [hunk 2]
```diff
@@ -9,18 +9,32 @@ module load cudatoolkit
...
+module load cray-python/3.9.4.2
+
+if [ -d "$HOME/sw/perlmutter/venvs/warpx" ]
+then
+  source $HOME/sw/perlmutter/venvs/warpx/bin/activate
+fi
+
+# an alias to request an interactive batch node for one hour
+#   for parallel execution, start on the batch node: srun <command>
+alias getNode="salloc -N 1 --ntasks-per-node=4 -t 1:00:00 -q interactive -C gpu -c 32 -G 4 -A $proj"
```
Comment (roelof-groenewald) [inline Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example]: Should this also include `--gpu-bind=single:1`? And same below?

Comment (ax3l) [reply Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example]: That's a good idea, yes.
```suggestion
alias getNode="salloc -N 1 --ntasks-per-node=4 -t 1:00:00 -q interactive -C gpu --gpu-bind=single:1 -c 32 -G 4 -A $proj"
```

File: Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example [hunk 3]
```diff
@@ -9,18 +9,32 @@ module load cudatoolkit
...
+# an alias to run a command on a batch node for up to 30min
+#   usage: runNode <command>
+alias runNode="srun -N 1 --ntasks-per-node=4 -t 0:30:00 -q interactive -C gpu -c 32 -G 4 -A $proj"
```
Comment (ax3l) [inline Tools/machines/perlmutter-nersc/perlmutter_warpx.profile.example]:
```suggestion
alias runNode="srun -N 1 --ntasks-per-node=4 -t 0:30:00 -q interactive -C gpu --gpu-bind=single:1 -c 32 -G 4 -A $proj"
```

**Comment by ax3l** [comment]
@roelof-groenewald do you still have the GPU visibility warnings with the setup in this PR?

**Comment by roelof-groenewald** [comment]
> @roelof-groenewald do you still have the GPU visibility warnings with the setup in this PR?

When I run on a full node (i.e. 4 GPUs) I don't get the GPU visibility warnings any more (nor the logging issue we discussed before). So thank you for these setup instructions!
However, I do now run into trouble when running on less than a full node. For example, when just requesting 2 GPUs (via `#SBATCH --ntasks-per-node=2`) I get the following error:
```
slurmstepd: error: Bind request 3 (0x8) does not specify any devices within the 
allocation. Binding to the first device in the allocation instead.
slurmstepd: error: Bind request 2 (0x4) does not specify any devices within the 
allocation. Binding to the first device in the allocation instead.
```
and the WarpX launch header:
```
Initializing CUDA...
CUDA initialized with 1 GPU(s) and 2 ranks.
MPI initialized with 2 MPI processes
```
Have you seen this behavior before? 
@PhilMiller suggested that I try setting the environment variable `CUDA_VISIBLE_DEVICES=0,1` but I haven't had a chance to try that yet.

**Comment by ax3l** [comment]
We can add this to `warpx.profile`, too:
```
export GPUS_PER_SOCKET=1
export GPUS_PER_NODE=4
```
but let us check if Perlmutter allows allocating non-node-exclusive jobs.

Yes, if we use slurm right (and it's properly configured) then `CUDA_VISIBLE_DEVICES` should be set automatically.

**Comment by ax3l** [comment]
@roelof-groenewald: @kngott just commented that there are no shared node allocations at the time at Perlmutter. So even if you only ask for less than 4 GPUs, you will block and be accounted (well, not bad at the moment :D) for the full node.

So in case you want to manually place multiple jobs on a node, you would still allocate the full node and then do
```bash
export CUDA_VISIBLE_DEVICES=0
srun ... &
export CUDA_VISIBLE_DEVICES=1
srun ... &
export CUDA_VISIBLE_DEVICES=2
srun ... &
export CUDA_VISIBLE_DEVICES=3
srun ... &
wait
```

**Comment by roelof-groenewald** [comment]
Sounds good. Thanks @ax3l and @kngott! 
So even in the case that I don't specifically want to run multiple jobs on a single node but want to run a job with 2 GPUs (for performance benchmarking for example), I will add `export CUDA_VISIBLE_DEVICES=0,1` to my job submission script.

**Review by roelof-groenewald — APPROVED** [review]

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**